### ISSUE TITLE
UnderlineNav v2 item can enter a selected non-current state

### ISSUE THREAD

**Issue by pveierland** [primer/react#2865]
### Description

During use of the UnderlineNav v2 component, it appears that the `UnderlineNav.Item` can end up showing as being selected, while its prop `aria-current` is set to `undefined` or `false`.

This may be caused by a race condition, as it seems to be triggered when moving between pages quickly / when rendering has not been completed.

Example of "Meta" not being selected and "Inputs" being selected:

![image](https://user-images.githubusercontent.com/5587375/216877703-af4ecb46-e536-45ff-96c1-9f96e261fc30.png)

Props showing that "Meta" item has `aria-current` set to `page` and should be selected:

![image](https://user-images.githubusercontent.com/5587375/216877854-b311da13-df74-4064-bc27-958c357d034c.png)

Props showing that "Inputs" item has `aria-current` set to `false` and should not be selected:

![image](https://user-images.githubusercontent.com/5587375/216877937-ba4581f0-0d09-4d24-89e3-95a03b598897.png)

Testing performed using React Primer v35.19.0 and Next.js v13.1.6 with wrapper component as detailed in: https://primer.style/react/drafts/UnderlineNav2#with-nextjs

### Steps to reproduce

1. Create page using https://primer.style/react/drafts/UnderlineNav2#with-nextjs template, with demanding components on each page being navigated between.
2. Navigate between pages using a separate navigation mechanism (not by clicking `UnderlineNav.Item`).
3. Sometimes the `UnderlineNav.Item` will end up in the state where it is shown visually as being selected, while having `aria-current` set to `undefined` or `false`.

Expected behavior would be for the `aria-current` and visual state of the `UnderlineNav.Item` to always be in sync.

### Version

v35.19.0

### Browser

Chrome, Firefox

**Comment by lesliecdubs** [comment]
@broccolinisoup do you want to take a look at this one?

**Comment by broccolinisoup** [comment]
I'm on it!

**Comment by pveierland** [comment]
Just noting a possibly related issue with the same component. Sometimes, not all items will be shown during a render (e.g. the last element might be missing), however when resizing the page the last element(s) will also be rendered. To be specific, it is not that the element is being shown in an extensible menu, it is that sometimes the last elements are not rendered at all.

**Comment by broccolinisoup** [comment]
@pveierland That is strange. "Sometimes" issues are tricky! -  Thanks so much for letting us know. I plan to look into this issue next week and hopefully resolve 🤞🏻

**Comment by pveierland** [comment]
Just adding a screenshot of how it appeared now. In this case the final item "Settings" was not added at all, while "Resources" and "Versions" are shown in an extension menu, even though there is available layout space for all items without using the dropdown.

![image](https://user-images.githubusercontent.com/5587375/225553739-1c79ce2e-c4cc-403f-afc4-cc4c6b235336.png)

Resizing the page makes all items render:

![image](https://user-images.githubusercontent.com/5587375/225554070-e30d12b0-8bab-470c-a53a-0ba4260ce59a.png)

**Comment by pveierland** [comment]
That does explain the behavior of the first render. The "Settings" entry also requires flag which is not set during the server render, and will only render when this flag is set on the client-side. However, wouldn't the component be re-rendered automatically after hydration and when these props have changed?

**Comment by broccolinisoup** [comment]
Hey @pveierland 👋🏻 Sorry for the late response!

>The "Settings" entry also requires a flag which is not set during the server render, and will only render when this flag is set on the client-side. 

Thanks for letting me know about that!

>However, wouldn't the component be re-rendered automatically after hydration and when these props have changed?

The component is re-rendered when the props are changed as well as when the viewport size is changed. However, regarding re-rendering automatically after hydration, I would say it is a option that needs to be chosen carefully as it will most likely cause a layout shift.

To my understanding and experience, the behaviour of a component both on the client and the server should be intentional. They can be different and that is completely fine **as long as it is intentional by design**. This is the case for UnderlineNav. The component relies on viewport size for accurate rendering on the client and for server, the last two item being hiding in the overflow menu in the first load is fine and we don't consider this as a bug. 

Let us know if you have further questions 🙌🏻

**Comment by broccolinisoup** [comment]
@pveierland Regarding the original issue, thank you so much for reporting it again. I have found a few issues related to keeping the selected item visible in SSR and I believe the original issue you mentioned is something to do with it. I'll post my findings here in the next couple of days hopefully and come up with more SSR compatible solution.

**Comment by peterbe** [comment]
> but that feels hacky :)

and obviously wouldn't work because a DOM click will trigger the underlying `onSelect` handler.

**Comment by broccolinisoup** [comment]
Hi @peterbe 👋🏻 Thanks for your comment, I can confirm now that indeed there is a bug here :) 
The orange underline is managed internally by a state called `selectedLink` and what I realised is that in some cases (like described above), `aria-current` and the `selectedLink` gets out of sync. I half way solved the issue but wanted to look into a little deeper. I prioritised this issue to unblock you and the original reporter of the issue (apologise for @pveierland for a late response), hopefully I'll be pushing a fix for this in the next few days. 

Thank you for taking the time to post reproducible info and screen recordings. I'll be coming here and update the issue once I resolve the problem. Thanks 🙏🏻

**Comment by broccolinisoup** [comment]
Hi 👋🏻 Thanks for your patience with me 🙏🏻 I think I got some good progress. I tested the scenarios on a sample SPA and seems like it is working. I still need to work through some styling and some bits and pieces - because I went ahead and did some refactoring - however I believe it is at a good stage to test out. 

I pushed a PR with my changes and every PR publishes a canary build https://github.com/primer/react/pull/3361 - would that be possible to test out the `@primer/react@0.0.0-20230605081804` version and let me know if you have any issues/feedback/concern etc? I thought it would be good to check in before going too far. Let me know what you think! Looking forward to hearing from you 🙌🏻 

cc @peterbe @pveierland

**Comment by peterbe** [comment]
> Hi 👋🏻 Thanks for your patience with me 🙏🏻 I think I got some good progress. I tested the scenarios on a sample SPA and seems like it is working. I still need to work through some styling and some bits and pieces - because I went ahead and did some refactoring - however I believe it is at a good stage to test out.
> 
> I pushed a PR with my changes and every PR publishes a canary build #3361 - would that be possible to test out the `@primer/react@0.0.0-20230605081804` version and let me know if you have any issues/feedback/concern etc? I thought it would be good to check in before going too far. Let me know what you think! Looking forward to hearing from you 🙌🏻
> 
> cc @peterbe @pveierland

I have not had a chance to fully try it yet. 
I got half-way. Distracted when working to get back into it. Will try again next week.

**Comment by peterbe** [comment]
@broccolinisoup Good news! Your canary build solves the problem. 
You can test on https://docs-internal-37792-34c9cc.preview.ghdocs.com/en/authentication/connecting-to-github-with-ssh/adding-a-new-ssh-key-to-your-github-account 
Works with mouse and with keyboard.

**Comment by broccolinisoup** [comment]
@peterbe Great news! Thanks so much for testing and giving me feedback. I will work on the final details and getting the PR ready for review, hopefully we will release it soon!

**Comment by echarles** [comment]
Hi, I struggled with the exact same issue while trying to control the selected tab from a global state. I have tried @primer/react@0.0.0-20230605081804 which indeed fixed the selected tab, but the navbar renders in stretched mode, and I am not able to make it take the full horizontal space.

<img width="180" alt="Screenshot 2023-06-22 at 12 47 29" src="https://github.com/primer/react/assets/226720/3d664b31-8477-4ca3-bfb5-95524b7b06cd">

**Comment by broccolinisoup** [comment]
hey @echarles I am sorry I missed this comment! I was aware of this overflowing styling issue. I finally managed to clean up [my PR](https://github.com/primer/react/pull/3361) and open for review and it should fix that styling too. 

@peterbe Sorry this took me so long! Hopefully I will manage to get it merged soon 🤞🏻
